content_script.js

```diff
@@ -1,82 +1,213 @@
+     * @const
+     */
+    this.FILTERED_TAGS = new Set(['script', 'noscript', 'style', 'link']);
+
+    /**
+     * @public {Array<string>} This array represents the serialized html that makes up an element or document. 
+     */
+    this.html = [];
+
+    /**
+     * @public {Object<number, string>} The keys represent an index in this.html.  The value
+     *    is a url at which the resource that belongs at that index can be retrieved.  The resource will eventually
+     *    be converted to a data url.  Because any given document being serialized could be an iframe which is
+     *    nested 1 or more levels into the root document, the exact quotes that will be used to surround the data url
+     *    must be determined when the frames are being put together, so that they can be properly escaped.  As a
+     *    result, this.html has a filler index both immediately before, and immediately after where the data url will
+     *    be placed (3 filler indices in total).
+     */
+    this.srcHoles = {};
+
+    /**
+     * @public {Object<number, string>} The keys represent an index in this.html.  The value
+     *    is a string that uniquely identifies an iframe, the serialized contents of which should be placed at that
+     *    index of this.html.
+     */
+    this.frameHoles = {};
+
+    /**
+     * @public {Array<number>} Each number in this.styleIndices corresponds to an index in this.html at which a
+     *    serialized style attribute is located. This is because there are styles that contain quotation marks.
+     *    Because any given document being serialized could be an iframe which is nested 1 or more levels into
+     *    the root document, the exact quotes that will be used must be determined when the frames are being put
+     *    together, so that they can be properly escaped.
+     */
+    this.styleIndices = [];
   }
-  else {
-    html.push(new Array(depth+1).join("  "));
-    html.push("<" + element.tagName + " ");
-
-    var style = element.ownerDocument.defaultView.getComputedStyle(element, null).cssText;
-    style = style.replace(/"/g, "'");
-    html.push('style="' + style + '" ');
-
-    var attributes = element.attributes;
-    if (attributes) {
-      for (var i = 0; i < attributes.length; i++) {
-        var attribute = attributes[i];
-        if (attribute.name.toLowerCase() == "src") {
-          html.push(attribute.name + "=");
-          holes[html.length] = attribute.value; // mark location to insert url later
-          html.push(""); // create filler entry
+
+  /**
+   * Takes an html element, and populates this object's fields such that it can eventually be converted into an
+   * html text file.
+   *
+   * @param {Element} element The Element to serialize
+   * @param {number} depth The number of parents this element has in the current frame
+   * @private
+   */ 
+  serializeTree(element, depth) {
```

Comment:
Maybe processTree is a better name, since this method builds the data structures from which the serialization can 
be generated, rather than actually serializing.